launch/launch/launch_service.py

```diff
@@ -207,187 +315,112 @@ def _is_idle(self):
+                                entity_futures,
+                                loop=this_loop,
+                                timeout=1.0,
+                                return_when=asyncio.FIRST_COMPLETED
+                            )
+                            if not done:
+                                self.__logger.debug(
+                                    'still waiting on futures: {}'.format(entity_futures)
+                                )
+                    else:
+                        await process_one_event_task
                 except KeyboardInterrupt:
-                    pass
+                    continue
                 except asyncio.CancelledError:
-                    self.__logger.error('asyncio run loop was canceled')
+                    self.__logger.error('run task was canceled')
+                    return_code = 1
+                    break
                 except Exception as exc:
                     msg = 'Caught exception in launch (see debug for traceback): {}'.format(exc)
                     self.__logger.debug(traceback.format_exc())
                     self.__logger.error(msg)
-                    self.__return_code = 1
-                    ret = self._shutdown(reason=msg, due_to_sigint=False, force_sync=True)
+                    ret = await self._shutdown(reason=msg, due_to_sigint=False)
                     assert ret is None, ret
-                    # restart run loop to let it shutdown properly
-                    run_loop_task = self.__loop_from_run_thread.create_task(self.__run_loop())
-        finally:
-            # No matter what happens, unset the loop and set running to false.
-            with self.__loop_from_run_thread_lock:
-                self.__shutting_down = False
-                self.__loop_from_run_thread = None
-                self.__context._set_asyncio_loop(None)
+                    return_code = 1
+                    # keep running to let things shutdown properly
+                    continue
+            return return_code
 
-            # Unset the signal handlers while not running.
-            on_sigint(None)
-            on_sigterm(None)
-            on_sigquit(None)
+    def run(self, *, shutdown_when_idle=True) -> int:
+        """
+        Run an event loop and visit all entities of all included LaunchDescription instances.
 
-            with self.__running_lock:
-                self.__running = False
+        This should only ever be run from the main thread and not concurrently with
+        asynchronous runs (see `run_async()` documentation).
 
-        return self.__return_code
+        :param: shutdown_when_idle if True (default), the service will shutdown when idle
+        """
+        # Make sure this has not been called from any thread but the main thread.
+        if threading.current_thread() is not threading.main_thread():
+            raise RuntimeError(
+                'LaunchService can only be run in the main thread.'
+            )
```

Comment:
run_async is already doing this check.

Reply:
True that. Removed in 9b16ec3c6b3214f307b162e297f64426341d842a.